File: clients/Go/5.1 mcp_edit_user/MCPEditUserClient.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net"
	"os"
)

type EditUserPayload struct {
	Command   string                 `json:"command"`
	Token     string                 `json:"token"`
	Arguments map[string]interface{} `json:"arguments"`
}

func sendEditUserRequest(serverIP string, serverPort int, token string, arguments map[string]interface{}) (string, error) {
	// Prepare the request payload
	payload := EditUserPayload{
		Command:   "edit_user",
		Token:     token,
		Arguments: arguments,
	}

	// Convert the payload to JSON
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	// Create a connection to the server
	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", serverIP, serverPort))
	if err != nil {
		return "", err
	}
	defer conn.Close()

	// Send the request
	_, err = conn.Write(payloadJSON)
	if err != nil {
		return "", err
	}

	// Receive the response
	var responseBuffer bytes.Buffer
	buf := make([]byte, 4096)
	for {
		n, err := conn.Read(buf)
		if err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return "", err
		}
		responseBuffer.Write(buf[:n])
		if n < 4096 {
			break
		}
	}

	return responseBuffer.String(), nil
}

func main() {
	serverIP := flag.String("server-ip", "", "IP address of the MCP server")
	serverPort := flag.Int("server-port", 0, "Port number of the MCP server")
	token := flag.String("token", "", "Authentication token")
	name := flag.String("name", "", "New name of the user")
	email := flag.String("email", "", "New email of the user")
	password := flag.String("password", "", "New password of the user")
	language := flag.String("language", "", "Preferred language of the user")
	timezone := flag.String("timezone", "", "Timezone of the user")
	roles := flag.String("roles", "", "Comma-separated list of roles to assign to the user")
	groups := flag.String("groups", "", "Comma-separated list of groups to assign to the user")
	usePublic := flag.Bool("usePublic", false, "Enable public knowledge base access")
	activateFtp := flag.Bool("activateFtp", false, "Activate FTP access")
	ftpPassword := flag.String("ftpPassword", "", "Password for FTP access")

	flag.Parse()

	if *serverIP == "" || *serverPort == 0 || *token == "" {
		fmt.Println("❌ ERROR: Required flags are missing.")
		flag.Usage()
		os.Exit(1)
	}

	// Prepare arguments map
	arguments := map[string]interface{}{}
	if *name != "" {
		arguments["name"] = *name
	}
	if *email != "" {
		arguments["email"] = *email
	}
	if *password != "" {
		arguments["password"] = *password
	}
	if *language != "" {
		arguments["language"] = *language
	}
	if *timezone != "" {
		arguments["timezone"] = *timezone
	}
	if *roles != "" {
		arguments["roles"] = *roles
	}
	if *groups != "" {
		arguments["groups"] = *groups
	}
	arguments["usePublic"] = *usePublic
	arguments["activateFtp"] = *activateFtp
	if *ftpPassword != "" {
		arguments["ftpPassword"] = *ftpPassword
	}

	// Send the request to edit the user
	fmt.Println("📤 Sending request to edit user...")
	response, err := sendEditUserRequest(*serverIP, *serverPort, *token, arguments)
	if err != nil {
		fmt.Printf("❌ ERROR: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("✔️ Response from server:")
	fmt.Println(response)
}

File: clients/Python/5.1 mcp_edit_user/MCPEditUserClient.py
import socket
import ssl
import json
import argparse
import sys

def send_edit_user_request(server_ip, server_port, token, name=None, email=None, password=None, language=None,
                           timezone=None, roles=None, groups=None, use_public=False, activate_ftp=False, ftp_password=None,
                           use_ssl=True, accept_self_signed=False):
    """
    Sends a request to edit an existing user to the MCP server.
    """
    payload = {
        "command": "edit_user",
        "token": token,
        "arguments": {
            "name": name,
            "email": email,
            "password": password,
            "language": language,
            "timezone": timezone,
            "roles": roles or [],
            "groups": groups or [],
            "usePublic": use_public,
            "activateFtp": activate_ftp,
            "ftpPassword": ftp_password
        }
    }

    # Remove keys with None values
    payload["arguments"] = {k: v for k, v in payload["arguments"].items() if v is not None}

    payload_json = json.dumps(payload)
    
    raw_socket = None
    client_socket = None

    try:
        raw_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        raw_socket.settimeout(10)

        if use_ssl:
            context = ssl.create_default_context()
            if accept_self_signed:
                context.check_hostname = False
                context.verify_mode = ssl.CERT_NONE
            client_socket = context.wrap_socket(raw_socket, server_hostname=server_ip)
        else:
            client_socket = raw_socket
        
        client_socket.connect((server_ip, server_port))
        client_socket.sendall(payload_json.encode('utf-8'))

        response = b""
        while True:
            part = client_socket.recv(4096)
            if not part:
                break
            response += part

        return response.decode('utf-8')
    except ssl.SSLError:
        return "Error: Server and/or client may require TLS encryption. Please enable SSL/TLS."
    except Exception as e:
        return f"Error: {e}"
    
    finally:
        if client_socket is not None:
            try:
                client_socket.shutdown(socket.SHUT_RDWR)
            except:
                pass
            client_socket.close()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Send a request to edit an existing user to the MCP server.",
        formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument("--server-ip", required=True, help="IP address of the MCP server")
    parser.add_argument("--server-port", required=True, type=int, help="Port number of the MCP server")
    parser.add_argument("--token", required=True, help="Authentication token")
    parser.add_argument("--name", help="New name of the user")
    parser.add_argument("--email", help="New email of the user")
    parser.add_argument("--password", help="New password of the user")
    parser.add_argument("--language", help="Preferred language of the user")
    parser.add_argument("--timezone", help="Timezone of the user")
    parser.add_argument("--roles", nargs="*", help="List of roles to assign to the user")
    parser.add_argument("--groups", nargs="*", help="List of groups to assign to the user")
    parser.add_argument("--usePublic", action="store_true", help="Enable public knowledge base access")
    parser.add_argument("--activateFtp", action="store_true", help="Activate FTP access")
    parser.add_argument("--ftpPassword", help="Password for FTP access")
    parser.add_argument("--use-ssl", action="store_true", help="Connect using SSL/TLS")
    parser.add_argument("--accept-self-signed", action="store_true", help="Accept self-signed certificates (disable certificate verification)")

    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)

    args = parser.parse_args()

    response = send_edit_user_request(
        args.server_ip,
        args.server_port,
        args.token,
        name=args.name,
        email=args.email,
        password=args.password,
        language=args.language,
        timezone=args.timezone,
        roles=args.roles,
        groups=args.groups,
        use_public=args.usePublic,
        activate_ftp=args.activateFtp,
        ftp_password=args.ftpPassword,
        use_ssl=args.use_ssl,
        accept_self_signed=args.accept_self_signed
    )
    print("Response from server:", response)
